ocaml/libs/clock/timer.ml

```diff
@@ -0,0 +1,53 @@
+type t = {start: Ptime.t; elapsed: Mtime_clock.counter; timeout: Mtime.Span.t}
+
+type remaining = Spare of Mtime.Span.t | Excess of Mtime.Span.t
+
+let is_shorter a ~than:b = Mtime.Span.compare a b < 0
+
+let is_longer a ~than:b = Mtime.Span.compare a b > 0
+
+let start ~timeout =
+  {start= Ptime_clock.now (); elapsed= Mtime_clock.counter (); timeout}
+
+let timeout {timeout; _} = timeout
+
+let elapsed t = Mtime_clock.count t.elapsed
+
+let remaining t =
+  let elapsed = Mtime_clock.count t.elapsed in
+  let difference = Mtime.Span.abs_diff t.timeout elapsed in
+  if Mtime.Span.compare t.timeout elapsed > 0 then
+    Spare difference
+  else
+    Excess difference
+
+let expired t = match remaining t with Spare _ -> false | Excess _ -> true
+
+let deadline_of t =
+  Mtime.Span.to_uint64_ns t.timeout
+  |> Int64.to_float
+  |> Ptime.Span.of_float_s
+  |> Option.get
+  |> Ptime.(Span.add Ptime.(to_span t.start))
+  |> Ptime.Span.to_float_s
+
+let shorten_by dur t =
+  let timeout =
+    if is_longer dur ~than:t.timeout then
+      Mtime.Span.zero
+    else
+      Mtime.Span.abs_diff dur t.timeout
+  in
+  {t with timeout}
```

Comment:
I'm not sure this is equivalent. If you have a timeout of 10s, and you are 8s in, and you shorten the timeout by 5s then you should get an excess of 3s, whereas instead you'd get a timer that has just expired.
Maybe that is intended, but we should document it.

Where is 'shorten_by' used?

Reply:
This function is the alternative to the `attempt_delay` parameter in your PR.

> shortening the timeout by a value larger than the current timeout seems wrong, and not sure what the expected semantics there would be

Semantically, when shortening a timeout to 0 (durations cannot be negative), a timer that always is expired is created. That is, at the time of creation, the timer had already expired.

This doesn't cause problems as the timers cannot go back in time and get their state inspected.

Reply:
Ah for `attempt_delay` all we need to know is whether the timer has expired, not by how much, so this is probably fine, if we explain it in the ocamldoc.
(The code in redo log attempts to take into account how much executing the operation would take, such that it completes the operation before the timeout, not merely start it)